### PR TITLE
chore(shopify): seperate pixel server side events logic from legacy tracker implementation

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## What are the changes introduced in this PR?

In order to offer a different way of how event properties are now totally conformant to the RudderStack Ecom Spec, and also support a `shopifyDetails` object inside context, on lines of how events from web pixel have been structured, the events from the new RudderStack Pixel App go through the destination transformations introduced in this PR.

## What is the related Linear task?

Resolves INT-2835

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here
These are introduced to separate pixel server side events logic from legacy tracker implementation, to facilitate more granular control over the server side events from the new RudderStack Pixel app on shopify. 

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

No existing change to the legacy tracker based app's destination transformation logic.

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

None

### Any technical or performance related pointers to consider with the change?

Some code duplication may occur to keep the transformed event same as earlier apart from 2 differences, in context and event properties as mentioned in the description.

@coderabbitai review

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [x] All changes manually tested?

- [x] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3849/test-report.html

File: src/v1/sources/shopify/transform.js [hunk 1]
```diff
@@ -13,7 +15,12 @@ const process = async (inputEvent) => {
     const responseV2 = await processEventFromPixel(event);
     return responseV2;
   }
-  // this is for common logic for server-side events processing for both pixel and tracker apps.
+  if (query_parameters && query_parameters?.version?.[0] === 'pixel') {
```
Comment (krishna2020) [inline src/v1/sources/shopify/transform.js:18]: are query_parameters values getting sent in array type from rudder-server ? cc: @vinayteki95

File: src/v1/sources/shopify/transform.js [hunk 2]
```diff
@@ -1,9 +1,11 @@
 /* eslint-disable @typescript-eslint/naming-convention */
-const { processEventFromPixel } = require('./pixelTransform');
+const { processEventFromPixel } = require('./webpixelTransformations/pixelTransform');
```
Comment (krishna2020) [inline src/v1/sources/shopify/transform.js:2]:
```suggestion
const { processPixelWebEvents } = require('./webpixelTransformations/pixelTransform');
```

File: src/v1/sources/shopify/transform.js [hunk 3]
```diff
@@ -1,9 +1,11 @@
...
 const { process: processWebhookEvents } = require('../../../v0/sources/shopify/transform');
+const { process: processWebhookEventsV2 } = require('./webhookTransformations/serverSideTransform');
 
 const process = async (inputEvent) => {
   const { event } = inputEvent;
+  const { query_parameters } = event;
   // check on the source Config to identify the event is from the tracker-based (legacy)
```
Comment (krishna2020) [inline src/v1/sources/shopify/transform.js:9]: remove this comment - no longer valid
```suggestion
  // check on the source Config to identify the event is from the tracker-based (legacy)
```

File: src/v1/sources/shopify/transform.js [hunk 4]
```diff
@@ -13,7 +15,12 @@ const process = async (inputEvent) => {
     const responseV2 = await processEventFromPixel(event);
```
Comment (krishna2020) [inline src/v1/sources/shopify/transform.js:15]:
```suggestion
    const pixelWebEventResponse = await processEventFromPixel(event);
```

File: src/v1/sources/shopify/transform.js [hunk 5]
```diff
@@ -13,7 +15,12 @@ const process = async (inputEvent) => {
...
     return responseV2;
   }
-  // this is for common logic for server-side events processing for both pixel and tracker apps.
+  if (query_parameters && query_parameters?.version?.[0] === 'pixel') {
+    // this is a server-side event from the webhook subscription made by the pixel app.
+    const responseV2 = await processWebhookEventsV2(event);
```
Comment (krishna2020) [inline src/v1/sources/shopify/transform.js:20]:
```suggestion
    const pixelWebhookEventResponse = await processWebhookEventsV2(event);
```

File: src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  const products = [];
+  lineItems.forEach((lineItem) => {
+    // const product = constructPayload(lineItem, lineItemsMappingJSON);
+    const product = constructPayload(lineItem, mapping);
+    products.push(product);
+  });
+  return products;
+};
+
+const createPropertiesForV2EcomEvent = (message) => {
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js:21]: what's v2 ecom event ?

File: src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js:1]: Add unit tests for this class

File: src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js (hunk 6 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js:21]: Updated the name to `createPropertiesForEcomEventFromWebhook`, it is the e-commerce event received from the pixel based app.

File: src/v1/sources/shopify/transform.js (hunk 5 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transform.js:20]: Updated

File: src/v1/sources/shopify/transform.js (hunk 4 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transform.js:15]: Addressed

File: src/v1/sources/shopify/transform.js (hunk 3 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transform.js:9]: thanks, updated it to describe the current flow

File: src/v1/sources/shopify/transform.js (hunk 2 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transform.js:2]: Updated

File: src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideUtlis.js:1]: Added

File: src/v1/sources/shopify/transform.js [hunk 7]
```diff
@@ -1,9 +1,11 @@
 /* eslint-disable @typescript-eslint/naming-convention */
-const { processEventFromPixel } = require('./pixelTransform');
+const { processEventFromPixel } = require('./webpixelTransformations/pixelTransform');
 const { process: processWebhookEvents } = require('../../../v0/sources/shopify/transform');
+const { process: processWebhookEventsV2 } = require('./webhookTransformations/serverSideTransform');
```
[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transform.js]: Somehow missed this, updated now

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:1]: Is `context.shopifyData` logic not handled for maintaining raw payload ?

Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:1]: It is `shopifyDetails` not `shopifyData`, same like we have for pixel events or in the custom pixel code.
[Reference](https://github.com/rudderlabs/rudder-transformer/blob/031f34fbafe61bfc224630726e91673cc58a2d64/src/v1/sources/shopify/webhookTransformations/serverSideTransform.js#L171)

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    }
+  }
+  message.setProperty(`integrations.${INTEGERATION}`, true);
+  message.setProperty('context.library', {
+    eventOrigin: 'server',
+    name: 'RudderStack Shopify Cloud',
+    version: '2.0.0',
+  });
+  message.setProperty('context.topic', shopifyTopic);
+  // attaching cart, checkout and order tokens in context object
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:167]: what's the use-case for setting tokens in context ?

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    case IDENTIFY_TOPICS.CUSTOMERS_UPDATE:
+      message = identifyPayloadBuilder(event);
+      break;
+    case ECOM_TOPICS.ORDERS_CREATE:
+    case ECOM_TOPICS.ORDERS_UPDATE:
+    case ECOM_TOPICS.CHECKOUTS_CREATE:
+    case ECOM_TOPICS.CHECKOUTS_UPDATE:
+      message = ecomPayloadBuilder(event, shopifyTopic);
+      break;
+    case 'carts_update':
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:113]: Is this event name still valid in pixel based flow ?

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    eventOrigin: 'server',
+    name: 'RudderStack Shopify Cloud',
+    version: '2.0.0',
+  });
+  message.setProperty('context.topic', shopifyTopic);
+  // attaching cart, checkout and order tokens in context object
+  message.setProperty(`context.cart_token`, event.cart_token);
+  message.setProperty(`context.checkout_token`, event.checkout_token);
+  // raw shopify payload passed inside context object under shopifyDetails
+  message.setProperty('context.shopifyDetails', event);
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:171]: raw event name from shopify should also be maintained in 'context.shopifyDetails'

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 9 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:109]: The event name is valid but the logic related to redis is not needed with the cart update event. 
Removed it in a followup commit

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 8 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:151]: It was added as per CS request for analytics purposes on warehouse. Kept in the pixel implementation too to maintain parity

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 9 above)
Comment (krishna2020) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:109]: refer `carts_update` from ECOM_TOPICS enum. doesn't this fall under ecomm spec ?

**Review by krishna2020 — APPROVED** [review]

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 9 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js]: Done.

**Review by utsabc — APPROVED** [review]

**Review by krishna2020 — APPROVED** [review]

**Pull request merged**